tools/IoTEdgeDevOps/DevOpsLib/AzureActiveDirectory.cs

```diff
@@ -0,0 +1,112 @@
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientId, nameof(azureActiveDirClientId));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientSecret, nameof(azureActiveDirClientSecret));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureResource, nameof(azureResource));
+
+            try
+            {
+                if ((DateTime.Compare(DateTime.UtcNow, AzureActiveDirectory.accessTokenExpiration) >= 0) ||
+                   (AzureActiveDirectory.accessToken == null) ||
+                   (AzureActiveDirectory.azureResource != azureResource))
+                {
+                    string requestUri = $"https://login.microsoftonline.com/{azureActiveDirTenant}/oauth2/token";
+                    const string grantType = "client_credentials";
+
+                    var client = new HttpClient();
+                    client.BaseAddress = new Uri(requestUri);
+
+                    var requestBody = new List<KeyValuePair<string, string>>();
+                    requestBody.Add(new KeyValuePair<string, string>("client_id", azureActiveDirClientId));
+                    requestBody.Add(new KeyValuePair<string, string>("client_secret", azureActiveDirClientSecret));
+                    requestBody.Add(new KeyValuePair<string, string>("grant_type", grantType));
+                    requestBody.Add(new KeyValuePair<string, string>("resource", azureResource));
+
+                    HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, "");
+                    // By default, if FormUrlEncodedContent() is used, the "Content-Type" is set to "application/x-www-form-urlencoded"
+                    // request.Content.Headers = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
+                    request.Content = new FormUrlEncodedContent(requestBody);
+
+                    var response = await client.SendAsync(request).ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+                    var responseMsg = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+                    AzureActiveDirectory.azureActiveDirTenant = azureActiveDirTenant;
+                    AzureActiveDirectory.azureActiveDirClientId = azureActiveDirClientId;
+                    AzureActiveDirectory.azureActiveDirClientSecret = azureActiveDirClientSecret;
+                    AzureActiveDirectory.azureResource = azureResource;
+
+                    var responseJson = JObject.Parse(responseMsg);
+                    AzureActiveDirectory.accessTokenExpiration = DateTime.UtcNow.AddSeconds((double)responseJson["expires_on"]);
+                    AzureActiveDirectory.accessToken = (string)responseJson["access_token"];
+                }
+
+                return AzureActiveDirectory.accessToken;
+            }
+            catch (Exception e)
+            {
+                throw e;
+            }
+        }
+
+        public Task<string> GetAccessToken() =>
```

Comment:
Why do you need this method?

Reply:
Just for the completeness that you can just get the token once it is set.